keylime/tpm/tpm2.py

```diff
@@ -1261,3 +1259,50 @@ def read_key_nvram(self):
             logger.debug("Invalid key length from NVRAM: %d" % (len(output)))
             return None
         return output
+
+    @staticmethod
+    def stringify_pcrs(log: dict) -> None:
+        '''Ensure that the PCR indices are strings
+
+        The YAML produced by `tpm2_eventlog`, when loaded by the yaml module,
+        uses integer keys in the dicts holding PCR contents.  That does not
+        correspond to any JSON data.  This method ensures those keys are
+        strings.
+        The log is untrusted because it ultimately comes from an untrusted
+        source and has been processed by software that has had bugs.'''
+        if (not isinstance(log, dict)) or 'pcrs' not in log:
+            return
+        old_pcrs = log['pcrs']
+        if not isinstance(old_pcrs, dict):
+            return
+        new_pcrs = dict()
+        for hash_alg, cells in old_pcrs.items():
+            if not isinstance(cells, dict):
+                new_pcrs[hash_alg] = cells
+                continue
+            new_pcrs[hash_alg] = {str(index): val for index, val in cells.items()}
+        log['pcrs'] = new_pcrs
+        return
+
+    def parse_binary_bootlog(self, log_bin:bytes) -> dict:
+        '''Parse and enrich a BIOS boot log
+
+        The input is the binary log.
+        The output is the result of parsing and applying other conveniences.'''
+        with tempfile.NamedTemporaryFile() as log_bin_file:
+            log_bin_file.write(log_bin)
+            log_bin_filename = log_bin_file.name
+            retDict = self.__run(['tpm2_eventlog', log_bin_filename])
+        log_parsed_strs = retDict['retout']
+        log_parsed_data = config.yaml_to_dict(log_parsed_strs, add_newlines=False)
```

Comment:
Since `config.yaml_to_dict` is not actually guaranteed to return a `dict`, there is a gap in the defensive programming here --- no checking that we actually got a `dict`.